futures-test/src/task/context.rs

```diff
@@ -0,0 +1,146 @@
+///
+/// let mut future = PinBox::new(async {
+///     let val = await!(rx1).unwrap();
+///     tx2.send(val).unwrap();
+/// });
+///
+/// let res = context.poll(future.as_pin_mut());
+/// assert_eq!(res, Poll::Pending);
+///
+/// assert_eq!(context.wake().count(), 0);
+/// tx1.send(5).unwrap();
+/// assert_eq!(context.wake().count(), 1);
+///
+/// let res = context.poll(future.as_pin_mut());
+/// assert_eq!(res, Poll::Ready(()));
+///
+/// assert_eq!(rx2.try_recv(), Ok(Some(5)));
+/// ```
+#[derive(Debug)]
+pub struct TestContext<W = wake::Panic, E = executor::Panic>
+where
+    W: Wake + 'static,
+    E: Executor,
+{
+    wake: Arc<W>,
+    executor: E,
+}
+
+impl<W, E> TestContext<W, E>
+where
+    W: Wake + 'static,
+    E: Executor,
+{
+    /// Create a new instance
+    pub fn new(wake: W, executor: E) -> Self {
+        Self {
+            wake: Arc::new(wake),
+            executor,
+        }
+    }
+
+    /// Get access to this [`TestContext`]s [`Wake`] implementation
+    pub fn wake(&self) -> &W {
+        &*self.wake
+    }
+
+    /// Get access to this [`TestContext`]s [`Executor`] implementation
+    pub fn executor(&self) -> &E {
+        &self.executor
+    }
+
+    /// Call a method with a [`Context`] referring to this [`TestContext`]'s
+    /// [`Wake`] and [`Executor`] implementations.
+    pub fn with<R, F>(&mut self, f: F) -> R
+    where
+        F: FnOnce(&mut Context) -> R,
+    {
+        let waker = unsafe { local_waker(self.wake.clone()) };
+        let mut context = Context::new(&waker, &mut self.executor);
+        f(&mut context)
+    }
+
+    /// Poll a [`Future`] once with a [`Context`] referring to this
+    /// [`TestContext`]'s [`Wake`] and [`Executor`] implementations.
+    pub fn poll<Fut>(&mut self, future: PinMut<Fut>) -> Poll<Fut::Output>
+    where
+        Fut: Future,
+    {
+        self.with(|cx| future.poll(cx))
+    }
+
+    /// Poll an [`Unpin`][Unpin] [`Future`][Future] once with a [`Context`] referring to this
+    /// [`TestContext`]'s [`Wake`] and [`Executor`] implementations.
+    pub fn poll_unpin<Fut>(&mut self, future: &mut Fut) -> Poll<Fut::Output>
+    where
+        Fut: Future + Unpin,
+    {
+        self.with(|cx| PinMut::new(future).poll(cx))
+    }
+}
+
+impl<W, E> Default for TestContext<W, E>
+where
+    W: Wake + Default + 'static,
+    E: Executor + Default,
+{
+    fn default() -> Self {
+        Self::new(W::default(), E::default())
+    }
+}
+
+/// Call a method with a [`Context`][futures_core::task::Context] using the
+/// specified [`Wake`][futures_core::task::Wake] and
+/// [`Executor`][futures_core::task::Executor] implementations.
+///
+/// # Examples
+///
+/// ```
+/// #![feature(async_await, await_macro, futures_api, pin)]
+///
+/// use futures::{future::Future, task::Poll};
+/// use futures_test::task::{self, wake, executor};
+/// use std::boxed::PinBox;
+///
+/// let mut future = PinBox::new(async {
+///     5
+/// });
```

Comment:
I suggest
```rust
let mut future = async { 5 };
pin_mut!(future);
```
I think we should prefer the `pin_mut!` macro in examples because users should reach for it first to avoid the heap allocation